mlir/tools/mlir-clang/Lib/clang-mlir.cc

```diff
@@ -770,6 +770,133 @@ MLIRScanner::VisitUnaryExprOrTypeTraitExpr(UnaryExprOrTypeTraitExpr *Uop) {
   }
 }
 
+bool isInAffineScope(Operation *op) {
+  auto *curOp = op;
+  while (auto *parentOp = curOp->getParentOp()) {
+    if (isa<mlir::AffineForOp>(parentOp))
+      return true;
+    curOp = parentOp;
+  }
+  return false;
+}
+
+bool hasAffineArith(Operation *op, AffineExpr &expr,
+                    mlir::Value &affineForIndVar) {
+  // skip IndexCastOp
+  if (isa<mlir::IndexCastOp>(op))
+    return hasAffineArith(op->getOperand(0).getDefiningOp(), expr,
+                          affineForIndVar);
+
+  // induction variable are modelled as memref<1xType>
+  // %1 = index_cast %induction : index to i32
+  // %2 = alloca() : memref<1xi32>
+  // store %1, %2[0] : memref<1xi32>
+  // ...
+  // %5 = load %2[0] : memref<1xf32>
+  if (isa<mlir::LoadOp>(op)) {
+    auto load = cast<mlir::LoadOp>(op);
+    auto loadOperand = load.getOperand(0);
+    if (loadOperand.getType().cast<MemRefType>().getShape().size() != 1)
+      return false;
+    auto maybeAllocaOp = loadOperand.getDefiningOp();
+    if (!isa<mlir::AllocaOp>(maybeAllocaOp))
+      return false;
+    auto allocaUsers = maybeAllocaOp->getUsers();
+    if (llvm::none_of(allocaUsers, [](mlir::Operation *op) {
+          if (isa<mlir::StoreOp>(op))
+            return true;
+          return false;
+        }))
+      return false;
+    for (auto user : allocaUsers)
+      if (auto storeOp = dyn_cast<mlir::StoreOp>(user)) {
+        auto storeOperand = storeOp.getOperand(0);
+        auto maybeIndexCast = storeOperand.getDefiningOp();
+        if (!isa<mlir::IndexCastOp>(maybeIndexCast))
+          return false;
+        auto indexCastOperand = maybeIndexCast->getOperand(0);
+        if (auto blockArg = indexCastOperand.dyn_cast<mlir::BlockArgument>()) {
+          if (auto affineForOp = dyn_cast<mlir::AffineForOp>(
+                  blockArg.getOwner()->getParentOp()))
+            affineForIndVar = affineForOp.getInductionVar();
+          else
+            return false;
+        }
+      }
+    return true;
+  }
+
+  // at this point we expect only AddIOp or MulIOp
+  if ((!isa<mlir::AddIOp>(op)) && (!isa<mlir::MulIOp>(op))) {
+    return false;
+  }
+
+  // make sure that the current op has at least one constant operand
+  // (ConstantIndexOp or ConstantIntOp)
+  if (llvm::none_of(op->getOperands(), [](mlir::Value operand) {
+        return (isa<mlir::ConstantIndexOp>(operand.getDefiningOp()) ||
```

Comment:
Not sure but should we check the generic `ConstantOp` as well, e.g., `isa<ConstantOp>(operand.getDefiningOp()) && operand.getType() == I32` (syntax might be wrong).

Reply:
Why so? There are 3 types of ```ConstantOp```: 1) ```ConstantIndexOp``` 2) ```ConstantIntOp``` and ```ConstantFloatOp```. I assume an induction variable to be of type 1 or 2.